From d8b7f916f6d34712b7853fe8224df4e114b42ac3 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Tue, 22 Apr 2025 15:17:54 +0200
Subject: [PATCH 0941/1010] evl/net: skb: save cached copy of netdev address

We need to keep a pointer to the originating device for the release
path (__free_evl_skb), since a regular protocol layer (e.g. UDP) might
use skb->dev_scratch which would cause skb->dev to be lost.

So we only rely on skb->dev as long as it is safe to do so, switching
to EVL_NET_CB(skb)->dev afterwards.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/skb.h  |  1 +
 kernel/evl/net/input.c |  7 ++++++-
 kernel/evl/net/skb.c   | 22 +++++++++++++++++-----
 3 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/include/evl/net/skb.h b/include/evl/net/skb.h
index a44c3a6829..ae6f091db1 100644
--- a/include/evl/net/skb.h
+++ b/include/evl/net/skb.h
@@ -22,6 +22,7 @@ struct iovec;
 struct evl_net_cb {
 	struct evl_net_handler *handler;
 	struct evl_socket *tracker;
+	struct net_device *dev;
 	union {
 		/* protocol-specific stuff should live here. */
 	};
diff --git a/kernel/evl/net/input.c b/kernel/evl/net/input.c
index 86e8ce25b4..61d3a11a40 100644
--- a/kernel/evl/net/input.c
+++ b/kernel/evl/net/input.c
@@ -167,12 +167,17 @@ EXPORT_SYMBOL_GPL(evl_net_wake_rx);
 void evl_net_receive(struct sk_buff *skb,
 		struct evl_net_handler *handler) /* in-band or oob */
 {
-	struct evl_netdev_state *est = skb->dev->oob_state.estate;
+	struct net_device *dev = skb->dev;
+	struct evl_netdev_state *est = dev->oob_state.estate;
+
+	if (EVL_WARN_ON(NET, dev == NULL))
+		return;
 
 	if (skb->next)
 		skb_list_del_init(skb);
 
 	EVL_NET_CB(skb)->handler = handler;
+	EVL_NET_CB(skb)->dev = dev;
 
 	/*
 	 * Enqueue the packet. The NIC driver is expected to call
diff --git a/kernel/evl/net/skb.c b/kernel/evl/net/skb.c
index 525323f308..a596cc9023 100644
--- a/kernel/evl/net/skb.c
+++ b/kernel/evl/net/skb.c
@@ -175,6 +175,15 @@ struct sk_buff *evl_net_dev_alloc_skb(struct net_device *dev,
 	 */
 	skb_reserve(skb, VLAN_HLEN);
 	skb->dev = real_dev;
+	/*
+	 * We need to keep a pointer to the originating device for the
+	 * release path (__free_evl_skb), since a protocol layer might
+	 * use skb->dev_scratch which would cause skb->dev to be
+	 * lost. So basically, we rely on skb->dev as long as it is
+	 * safe to do so, switching to EVL_NET_CB(skb)->dev
+	 * afterwards.
+	 */
+	EVL_NET_CB(skb)->dev = real_dev;
 
 	return skb;
 }
@@ -199,10 +208,9 @@ static void free_inband_skb(struct sk_buff *skb)
 	}
 }
 
-static void __free_evl_skb(struct sk_buff *skb)
+static void __free_evl_skb(struct sk_buff *skb, struct net_device *dev)
 {
 	struct skb_shared_info *shinfo = skb_shinfo(skb);
-	struct net_device *dev = skb->dev;
 	struct evl_netdev_state *est = dev->oob_state.estate;
 	unsigned long flags;
 
@@ -242,10 +250,14 @@ static void __free_evl_skb(struct sk_buff *skb)
 /*
  * Free an skb we originally allocated from our pool. The caller has
  * exclusive ownership on this (i.e. no other reference is pending).
+ *
+ * CAUTION: skb->dev might be invalid, always use the cached value in
+ * EVL_NET_CB(skb)->dev on the release path instead. See comment in
+ * evl_net_dev_alloc_skb().
  */
 static void free_evl_skb(struct sk_buff *skb)
 {
-	struct net_device *dev = skb->dev;
+	struct net_device *dev = EVL_NET_CB(skb)->dev;
 	struct sk_buff *fskb, *nskb;
 
 	if (EVL_WARN_ON(NET, dev == NULL))
@@ -258,13 +270,13 @@ static void free_evl_skb(struct sk_buff *skb)
 	for (fskb = skb_shinfo(skb)->frag_list; fskb; fskb = nskb) {
 		netdev_dbg(dev, "releasing frag %px from %px\n", fskb, skb);
 		nskb = fskb->next;
-		__free_evl_skb(fskb);
+		__free_evl_skb(fskb, dev);
 	}
 
 	netdev_dbg(dev, "releasing skb %px (has_frags=%d)\n",
 		skb, skb_has_frag_list(skb));
 
-	__free_evl_skb(skb);
+	__free_evl_skb(skb, dev);
 }
 
 static void __free_skb(struct sk_buff *skb)
-- 
2.30.2

